store/tikv/lock_resolver.go

```diff
@@ -357,44 +372,51 @@ func (t *txnExpireTime) value() int64 {
 // If the primary key is still locked, it will launch a Rollback to abort it.
 // To avoid unnecessarily aborting too many txns, it is wiser to wait a few
 // seconds before calling it after Prewrite.
-func (lr *LockResolver) GetTxnStatus(txnID uint64, primary []byte) (TxnStatus, error) {
+func (lr *LockResolver) GetTxnStatus(txnID uint64, callerStartTS uint64, primary []byte) (TxnStatus, error) {
 	var status TxnStatus
 	bo := NewBackoffer(context.Background(), cleanupMaxBackoff)
 	currentTS, err := lr.store.GetOracle().GetLowResolutionTimestamp(bo.ctx)
 	if err != nil {
 		return status, err
 	}
-	return lr.getTxnStatus(bo, txnID, primary, currentTS)
+	return lr.getTxnStatus(bo, txnID, primary, callerStartTS, currentTS)
 }
 
-func (lr *LockResolver) getTxnStatusFromLock(bo *Backoffer, l *Lock) (TxnStatus, error) {
+func (lr *LockResolver) getTxnStatusFromLock(bo *Backoffer, l *Lock, callerStartTS uint64, cleanTxns map[uint64]map[RegionVerID]struct{}) (TxnStatus, error) {
 	// NOTE: l.TTL = 0 is a special protocol!!!
 	// When the pessimistic txn prewrite meets locks of a txn, it should rollback that txn **unconditionally**.
-	// In this case, TiKV set the lock TTL = 0, and TiDB use currentTS = 0 to call
-	// getTxnStatus, and getTxnStatus with currentTS = 0 would rollback the transaction.
+	// In this case, TiKV use lock TTL = 0 to notify TiDB, and TiDB should resolve the lock!
 	if l.TTL == 0 {
-		return lr.getTxnStatus(bo, l.TxnID, l.Primary, 0)
+		var status TxnStatus
+		cleanRegions, exists := cleanTxns[l.TxnID]
+		if !exists {
+			cleanRegions = make(map[RegionVerID]struct{})
+			cleanTxns[l.TxnID] = cleanRegions
+		}
+		err := lr.resolveLock(bo, l, status, cleanRegions)
```

Comment:
Are you removing secondary locks regardless of the primary?

Reply:
I think it should resolve the primary lock first.

Reply:
It seems that the resovlveLock function didn't check the primary but only removes the specified lock. And by the way, the unit test didn't test if it works on a multi-key transaction's secondary key, so this mistake is not discovered by the unit test.

Reply:
If the primary lock's txn_size is greater than 16, `resolveLock` will resolve the whole region.

Reply:
> Are you removing secondary locks regardless of the primary?

No, the code is 
```
if status.ttl == 0 {
         err = lr.resolveLock(bo, l, status, cleanRegions)
```

status.ttl == 0 means the status of the primary lock is rollbacked / committed.

Reply:
I guess you misunderstand those two functions:

ResolveLocks  = getTxnStatus + resolveLock
resolveLock = send ResolveLockRequest request to TiKV

@MyonKeminta

Reply:
> resolveLock

So what?

Reply:
> If the primary lock's txn_size is greater than 16, `resolveLock` will resolve the whole region.

So what?

Reply:
Sorry, I misunderstood it before. Originally, `resolveLock` will get the primary lock's status first, but now it just does resolve. We don't know the transaction status if we meet a lock with ttl=0, so we can't roll back it.